### ISSUE TITLE
如何集成到swin-transformer仅增加少量参数?

### ISSUE THREAD

**Issue by waterweek** [Linwei-Chen/FDConv#10]
我将FDConv代替swin-transformer中MLP,参数量从原来3670MB激增到15154MB.代码如下:
`from .FDConv import FDConv
class MLPWithFDConv(nn.Module):
    def __init__(self, in_features, hidden_features, out_features=None, act_layer=nn.GELU, drop=0., input_resolution=None):
        super().__init__()
        out_features = out_features or in_features
        self.input_resolution = input_resolution
        self.fdconv1 = FDConv(in_features, hidden_features, kernel_size=1)
        self.act = act_layer()
        self.drop1 = nn.Dropout(drop)
        self.fdconv2 = FDConv(hidden_features, out_features, kernel_size=1)
        self.drop2 = nn.Dropout(drop)
        
    def forward(self, x):
        B, L, C = x.shape
        H, W = self.input_resolution
        assert L == H * W, "input feature has wrong size"
        # 转换维度以适配卷积操作
        x = x.permute(0, 2, 1).view(B, C, H, W)  # [B, C, H, W]
        x = self.fdconv1(x)
        x = self.act(x)
        x = self.drop1(x)
        x = self.fdconv2(x)
        x = x.view(B, C, H*W).permute(0, 2, 1)  # [B, L, C]
        x = self.drop2(x)
        return x
 self.mlp = MLPWithFDConvin_features=dim,hidden_features=mlp_hidden_dim,act_layer=act_layer,drop=drop,input_resolution=input_resolution)#mlp_hidden_dim = int(dim * mlp_ratio), mlp_ratio=4
`
我使用模型是swin_tiny_patch4_window7_224,
_C.MODEL.SWIN.PATCH_SIZE = 4
_C.MODEL.SWIN.IN_CHANS = 3
_C.MODEL.SWIN.EMBED_DIM = 96
_C.MODEL.SWIN.DEPTHS = [ 2, 2, 6, 2 ]
_C.MODEL.SWIN.NUM_HEADS = [ 3, 6, 12, 24 ]
_C.MODEL.SWIN.WINDOW_SIZE = 7
_C.MODEL.SWIN.MLP_RATIO = 4.

**Comment by Linwei-Chen** [comment]
感谢关注！FFN一般会有一个4x的扩大，所以FDConv中reduction=0.0625应该还需要/4

**Comment by waterweek** [comment]
谢谢回复!我尝试把FDConv.py文件中reduction=0.0625换成了reduction=0.015625,但参数量还是比较大15106MB.

**Comment by Linwei-Chen** [comment]
这里有一段vit的代码供参考
```
for _ in range(num_fcs - 1):
            layers.append(
                Sequential(
                    # Linear(in_channels, feedforward_channels), 
                    FDConv(
                        in_channels, 
                        feedforward_channels,
                        kernel_size=1,
                        reduction=0.0625, 
                        kernel_num=64,
                        use_fdconv_if_c_gt=16, #if channel greater or equal to 16, e.g., 64, 128, 256, 512
                        use_fdconv_if_k_in=[1, 3], #if kernel_size in the list
                        use_fbm_if_k_in=[1], #if kernel_size in the list
                        kernel_temp=1.0,
                        temp=1.0,
                        att_multi=2.0,
                        param_ratio=1,
                        param_reduction=1.0,
                        ksm_only_kernel_att=False,
                        att_grid=1,
                        use_ksm_local=True,
                        ksm_local_act='sigmoid',
                        spatial_freq_decompose=False,
                        convert_param=True,
                        linear_mode=False,
                        fbm_cfg={
                            'k_list':[2, 4, 8],
                            'lowfreq_att':False,
                            'fs_feat':'feat',
                            'act':'sigmoid',
                            'spatial':'conv',
                            'spatial_group':1,
                            'spatial_kernel':3,
                            'init':'zero',
                            'global_selection':False,
                        },
                    ),
                    self.activate,
                    nn.Dropout(ffn_drop)))
            in_channels = feedforward_channels
        layers.append(
            # Linear(feedforward_channels, embed_dims)
            FDConv(
                    feedforward_channels, 
                    embed_dims,
                    kernel_size=1,
                    reduction=1/64, 
                    kernel_num=64,
                    use_fdconv_if_c_gt=16, #if channel greater or equal to 16, e.g., 64, 128, 256, 512
                    use_fdconv_if_k_in=[1, 3], #if kernel_size in the list
                    use_fbm_if_k_in=[1], #if kernel_size in the list
                    kernel_temp=1.0,
                    temp=1.0,
                    att_multi=2.0,
                    param_ratio=1,
                    param_reduction=1.0,
                    ksm_only_kernel_att=False,
                    att_grid=1,
                    use_ksm_local=True,
                    ksm_local_act='sigmoid',
                    spatial_freq_decompose=False,
                    convert_param=True,
                    linear_mode=False,
                    fbm_cfg={
                        'k_list':[2, 4, 8],
                        'lowfreq_att':False,
                        'fs_feat':'feat',
                        'act':'sigmoid',
                        'spatial':'conv',
                        'spatial_group':1,
                        'spatial_kernel':3,
                        'init':'zero',
                        'global_selection':False,
                    },
                )
            )
```

**Comment by waterweek** [comment]
好的,我想问论文中swin+FDConv后参数量从48M->51M,FDConv是替换了swin-t所有阶段的MLP吗?还是有选择的替换?
